File: regions/system/library/typescript/building-blocks/src/inmemory_binding.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { InMemoryInputBinding, InMemoryOutputBinding } from './inmemory_binding.js';
import { ComponentError } from './errors.js';

// InMemoryInputBinding のテスト: インメモリキューからデータを読み取る入力バインディングの動作を検証する。
describe('InMemoryInputBinding', () => {
  let binding: InMemoryInputBinding;

  beforeEach(() => {
    binding = new InMemoryInputBinding();
  });

  it('初期状態は uninitialized', async () => {
    expect(await binding.status()).toBe('uninitialized');
  });

  it('init 後は ready になる', async () => {
    await binding.init();
    expect(await binding.status()).toBe('ready');
  });

  it('close 後は closed になりキューがクリアされる', async () => {
    await binding.init();
    binding.push({ data: new Uint8Array([1]), metadata: {} });
    await binding.close();
    expect(await binding.status()).toBe('closed');
  });

  it('デフォルト name は inmemory-input-binding', () => {
    expect(binding.name).toBe('inmemory-input-binding');
    expect(binding.componentType).toBe('binding.input');
  });

  it('コンストラクタで name を指定できる', () => {
    const named = new InMemoryInputBinding('custom-input');
    expect(named.name).toBe('custom-input');
  });

  it('metadata は backend=memory と direction=input を返す', () => {
    expect(binding.metadata()).toEqual({ backend: 'memory', direction: 'input' });
  });

  it('push したデータを read で取得できる（FIFO順）', async () => {
    await binding.init();
    binding.push({ data: new Uint8Array([1]), metadata: { seq: '1' } });
    binding.push({ data: new Uint8Array([2]), metadata: { seq: '2' } });

    const first = await binding.read();
    const second = await binding.read();

    expect(first.data).toEqual(new Uint8Array([1]));
    expect(first.metadata?.seq).toBe('1');
    expect(second.data).toEqual(new Uint8Array([2]));
  });

  it('キューが空のときに read すると ComponentError をスローする', async () => {
    await binding.init();
    await expect(binding.read()).rejects.toBeInstanceOf(ComponentError);
  });
});

// InMemoryOutputBinding のテスト: invoke の呼び出し履歴記録とモックレスポンス設定機能を検証する。
describe('InMemoryOutputBinding', () => {
  let binding: InMemoryOutputBinding;

  beforeEach(() => {
    binding = new InMemoryOutputBinding();
  });

  it('初期状態は uninitialized', async () => {
    expect(await binding.status()).toBe('uninitialized');
  });

  it('init 後は ready になる', async () => {
    await binding.init();
    expect(await binding.status()).toBe('ready');
  });

  it('close 後は closed になり呼び出し履歴がクリアされる', async () => {
    await binding.init();
    await binding.invoke('op', new Uint8Array([1]));
    await binding.close();
    expect(await binding.status()).toBe('closed');
    expect(binding.lastInvocation()).toBeUndefined();
  });

  it('デフォルト name は inmemory-output-binding', () => {
    expect(binding.name).toBe('inmemory-output-binding');
    expect(binding.componentType).toBe('binding.output');
  });

  it('コンストラクタで name を指定できる', () => {
    const named = new InMemoryOutputBinding('custom-output');
    expect(named.name).toBe('custom-output');
  });

  it('metadata は backend=memory と direction=output を返す', () => {
    expect(binding.metadata()).toEqual({ backend: 'memory', direction: 'output' });
  });

  it('invoke の前は lastInvocation が undefined', async () => {
    await binding.init();
    expect(binding.lastInvocation()).toBeUndefined();
  });

  it('invoke が呼び出し履歴を記録する', async () => {
    await binding.init();
    await binding.invoke('send', new Uint8Array([1, 2]), { key: 'val' });

    const inv = binding.lastInvocation();
    expect(inv).toBeDefined();
    expect(inv!.operation).toBe('send');
    expect(inv!.data).toEqual(new Uint8Array([1, 2]));
    expect(inv!.metadata).toEqual({ key: 'val' });
  });

  it('invoke はデフォルトで入力データをそのまま返す', async () => {
    await binding.init();
    const resp = await binding.invoke('op', new Uint8Array([42]));
    expect(resp.data).toEqual(new Uint8Array([42]));
  });

  it('setResponse でモックレスポンスを設定できる', async () => {
    await binding.init();
    binding.setResponse({ data: new Uint8Array([99]), metadata: {} });
    const resp = await binding.invoke('op', new Uint8Array([1]));
    expect(resp.data).toEqual(new Uint8Array([99]));
  });

  it('setResponse でモックエラーを設定できる', async () => {
    await binding.init();
    const mockError = new Error('invoke error');
    binding.setResponse(undefined, mockError);
    await expect(binding.invoke('op', new Uint8Array([1]))).rejects.toBe(mockError);
  });

  it('allInvocations で全履歴を取得できる', async () => {
    await binding.init();
    await binding.invoke('op1', new Uint8Array([1]));
    await binding.invoke('op2', new Uint8Array([2]));

    const all = binding.allInvocations();
    expect(all).toHaveLength(2);
    expect(all[0].operation).toBe('op1');
    expect(all[1].operation).toBe('op2');
  });

  it('reset で履歴とモック設定をクリアできる', async () => {
    await binding.init();
    binding.setResponse({ data: new Uint8Array([99]), metadata: {} });
    await binding.invoke('op', new Uint8Array([1]));

    binding.reset();

    expect(binding.lastInvocation()).toBeUndefined();
    // reset 後はデフォルト動作（入力をそのまま返す）に戻る
    const resp = await binding.invoke('op', new Uint8Array([7]));
    expect(resp.data).toEqual(new Uint8Array([7]));
  });
});

File: regions/system/library/go/building-blocks/inmemory_binding_test.go
package buildingblocks

import (
	"context"
	"errors"
	"testing"
)

func TestInMemoryOutputBinding_InitAndStatus(t *testing.T) {
	b := NewInMemoryOutputBinding()
	ctx := context.Background()

	if b.Status(ctx) != StatusUninitialized {
		t.Errorf("expected StatusUninitialized, got %s", b.Status(ctx))
	}
	if err := b.Init(ctx, Metadata{}); err != nil {
		t.Fatalf("Init failed: %v", err)
	}
	if b.Status(ctx) != StatusReady {
		t.Errorf("expected StatusReady, got %s", b.Status(ctx))
	}
}

func TestInMemoryOutputBinding_Name(t *testing.T) {
	b := NewInMemoryOutputBinding()
	if b.Name() != "inmemory-binding" {
		t.Errorf("unexpected Name: %q", b.Name())
	}
	if b.Version() != "1.0.0" {
		t.Errorf("unexpected Version: %q", b.Version())
	}
}

func TestInMemoryOutputBinding_InvokeRecords(t *testing.T) {
	b := NewInMemoryOutputBinding()
	ctx := context.Background()
	_ = b.Init(ctx, Metadata{})

	if b.LastInvocation() != nil {
		t.Error("expected nil LastInvocation before any Invoke")
	}

	resp, err := b.Invoke(ctx, "send", []byte("payload"), map[string]string{"key": "val"})
	if err != nil {
		t.Fatalf("Invoke failed: %v", err)
	}
	if resp == nil {
		t.Fatal("expected non-nil response")
	}

	inv := b.LastInvocation()
	if inv == nil {
		t.Fatal("expected LastInvocation to be set")
	}
	if inv.Operation != "send" {
		t.Errorf("expected Operation 'send', got %q", inv.Operation)
	}
	if string(inv.Data) != "payload" {
		t.Errorf("expected Data 'payload', got %q", inv.Data)
	}
	if inv.Metadata["key"] != "val" {
		t.Errorf("expected Metadata key 'val', got %q", inv.Metadata["key"])
	}
}

func TestInMemoryOutputBinding_SetResponse(t *testing.T) {
	b := NewInMemoryOutputBinding()
	ctx := context.Background()
	_ = b.Init(ctx, Metadata{})

	b.SetResponse(&BindingResponse{Data: []byte("ok")}, nil)

	resp, err := b.Invoke(ctx, "op", nil, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if string(resp.Data) != "ok" {
		t.Errorf("expected 'ok', got %q", resp.Data)
	}
}

func TestInMemoryOutputBinding_SetResponseError(t *testing.T) {
	b := NewInMemoryOutputBinding()
	ctx := context.Background()
	_ = b.Init(ctx, Metadata{})

	want := errors.New("invoke error")
	b.SetResponse(nil, want)

	_, err := b.Invoke(ctx, "op", nil, nil)
	if !errors.Is(err, want) {
		t.Errorf("expected %v, got %v", want, err)
	}
}

func TestInMemoryOutputBinding_Reset(t *testing.T) {
	b := NewInMemoryOutputBinding()
	ctx := context.Background()
	_ = b.Init(ctx, Metadata{})

	b.SetResponse(&BindingResponse{Data: []byte("x")}, nil)
	_, _ = b.Invoke(ctx, "op", nil, nil)

	b.Reset()

	if b.LastInvocation() != nil {
		t.Error("expected nil LastInvocation after Reset")
	}
	// After reset, default empty response is returned.
	resp, err := b.Invoke(ctx, "op2", nil, nil)
	if err != nil {
		t.Fatalf("unexpected error after Reset: %v", err)
	}
	if resp.Data != nil {
		t.Errorf("expected nil Data after Reset, got %v", resp.Data)
	}
}

func TestInMemoryOutputBinding_Close(t *testing.T) {
	b := NewInMemoryOutputBinding()
	ctx := context.Background()
	_ = b.Init(ctx, Metadata{})

	if err := b.Close(ctx); err != nil {
		t.Fatalf("Close failed: %v", err)
	}
	if b.Status(ctx) != StatusClosed {
		t.Errorf("expected StatusClosed, got %s", b.Status(ctx))
	}
}
